File: Lab2/lab2-go/internal/timer/timer.go
package timer

import (
	"conc-lab-2/internal/calculator"
	"time"
)

type timer struct {
	calc calculator.Calculator
}

func NewTimer(calc calculator.Calculator) calculator.Calculator {
	return &timer{
		calc: calc,
	}
}

func (t *timer) CalcSum(arr []int64) int64 {
	now := time.Now()
	t.calc.CalcSum(arr)
	after := time.Since(now)
	println("Time:", after.Milliseconds(), "ms")
	return arr[0]
}

File: Lab3/program/timer/timer.cpp
#include "timer.h"
using namespace chrono;

int64_t TimerDecorator::calcSum(int64_t *arr, int arr_size) {
  auto start = high_resolution_clock::now();

  int64_t result = wrapped->calcSum(arr, arr_size);

  auto end = high_resolution_clock::now();
  auto duration = duration_cast<milliseconds>(end - start).count();

  cout << "Time: " << duration << " ms" << endl;

  return result;
}
